### PR TITLE
Feature/shared model swig

### PR DESCRIPTION
**Pull request by luckychess**
## What is this pull request?
Swig bindings for Python and Java. Scope is shared model builders and signing built objects.
  
## How to use the features provided in the pull request?
Bindings libraries are disabled by default. To generate them add ```-DSWIG_PYTHON=ON``` and/or ```-DSWIG_JAVA=ON``` option to your cmake command.
Then you can build Iroha as usual. After the build process is completed navigate to ```build/shared_model//bindings/```. You can use Iroha crypto from Python in a next way:
```
import iroha

builder = iroha.ModelBuilder()
crypto = iroha.ModelCrypto()

me_kp = crypto.generateKeypair()
peer_kp = crypto.generateKeypair()
signatory_kp = crypto.generateKeypair()
account_kp = crypto.generateKeypair()
time = 1512549580
startCounter = 1
creator = "me"
signatory = "fyodorkek-san"

commands = []

commands.append(builder.txCounter(startCounter).creatorAccountId(creator).createdTime(time).addPeer("127.0.0.1:50051", peer_kp.publicKey()).build())

commands.append(builder.txCounter(startCounter+1).creatorAccountId(creator).createdTime(time+1).createDomain("iroha", "admin").build())

commands.append(builder.txCounter(startCounter+2).creatorAccountId(creator).createdTime(time+2).createAccount("luckychess", "iroha", account_kp.publicKey()).build())

commands.append(builder.txCounter(startCounter+3).creatorAccountId(creator).createdTime(time+3).addSignatory(signatory, signatory_kp.publicKey()).build())

commands.append(builder.txCounter(startCounter+4).creatorAccountId(creator).createdTime(time+4).removeSignatory(signatory, signatory_kp.publicKey()).build())

commands.append(builder.txCounter(startCounter+5).creatorAccountId(creator).createdTime(time+5).setAccountQuorum(signatory, 2).build())

commands.append(builder.txCounter(startCounter+6).creatorAccountId(creator).createdTime(time+6).assetQuantity("luckychess@iroha", "irohacoin#iroha", "123.45").build())

commands.append(builder.txCounter(startCounter+7).creatorAccountId(creator).createdTime(time+7).transferAsset("luckychess@iroha", "kamil@iroha", "irohacoin#iroha", "hello world", "12.45").build())


protoHelper = iroha.ModelTransactionProto()

for command in commands:
    print(protoHelper.signAndAddSignature(command, me_kp).toString())
```
And from Java:
```
import java.math.BigInteger;
import java.util.*;

public class my {
  static {
    try {
        System.loadLibrary("iroha");
    } catch (UnsatisfiedLinkError e) {
      System.err.println("Native code library failed to load. \n" + e);
      System.exit(1);
    }
  }

  public static void main(String argv[]) {
    ModelBuilder builder = new ModelBuilder();
    ModelCrypto crypto = new ModelCrypto();
    Keypair me_kp = crypto.generateKeypair();
    Keypair peer_kp = crypto.generateKeypair();
    Keypair signatory_kp = crypto.generateKeypair();
    Keypair account_kp = crypto.generateKeypair();
    long time = 1512549580;
    long startCounter = 1;
    String creator = "me";
    String signatory = "fyodorkek-san";

    ArrayList<UnsignedTx> commands = new ArrayList<UnsignedTx>();
    commands.add(builder.txCounter(BigInteger.valueOf(startCounter)).creatorAccountId(creator).createdTime(BigInteger.valueOf(time)).addPeer("127.0.0.1:50051", peer_kp.publicKey()).build());
    commands.add(builder.txCounter(BigInteger.valueOf(startCounter+1)).creatorAccountId(creator).createdTime(BigInteger.valueOf(time+1)).createDomain("iroha", "admin").build());
    commands.add(builder.txCounter(BigInteger.valueOf(startCounter+2)).creatorAccountId(creator).createdTime(BigInteger.valueOf(time+2)).createAccount("luckychess", "iroha", account_kp.publicKey()).build());
    commands.add(builder.txCounter(BigInteger.valueOf(startCounter+3)).creatorAccountId(creator).createdTime(BigInteger.valueOf(time+3)).addSignatory(signatory, signatory_kp.publicKey()).build());
    commands.add(builder.txCounter(BigInteger.valueOf(startCounter+4)).creatorAccountId(creator).createdTime(BigInteger.valueOf(time+4)).removeSignatory(signatory, signatory_kp.publicKey()).build());
    commands.add(builder.txCounter(BigInteger.valueOf(startCounter+5)).creatorAccountId(creator).createdTime(BigInteger.valueOf(time+5)).setAccountQuorum(signatory, 2).build());
    commands.add(builder.txCounter(BigInteger.valueOf(startCounter+6)).creatorAccountId(creator).createdTime(BigInteger.valueOf(time+6)).assetQuantity("luckychess@iroha", "irohacoin#iroha", "123.45").build());
    commands.add(builder.txCounter(BigInteger.valueOf(startCounter+7)).creatorAccountId(creator).createdTime(BigInteger.valueOf(time+7)).transferAsset("luckychess@iroha", "kamil@iroha", "irohacoin#iroha", "hello world", "12.45").build());

    ModelTransactionProto protoHelper = new ModelTransactionProto();
    for (UnsignedTx tx: commands) {
      System.out.println(protoHelper.signAndAddSignature(tx, me_kp).toString());
    }
  }
}
```
and then run
```
javac *.java
```
If everything has gone well you will get an output like:
```
Blob: [0a430a3512330a0f3132372e302e302e313a35303035311220d8401b24e7197f25c914094d810574620be36fb63a3e8f66bce7ff1120b7543912026d65180120ccd99ed10512640a20132f93747b3ad9dd91252b582aa22c5c1510de1d51778e0c9600e4208103201612400b3de8b76753750afb75a986f7757f1de09afd50ee08b0b2790bfc924da92161e4697a5de7e1beea9d020d0b58b477bfeaf221ba05eaa9e620a36b63b0b3410c ]
Blob: [0a1e0a10320e0a0569726f6861120561646d696e12026d65180220cdd99ed10512640a20132f93747b3ad9dd91252b582aa22c5c1510de1d51778e0c9600e4208103201612401bbb0d53f1aa9e8152de77a65b5aaceac8c5a69e4ab0c32ebf67fb328902be4e5b55917f543406906b1642be2333962f6de65a8c3ae16f70e88dfa8d19c5fb03 ]
Blob: [0a450a372a350a0a6c75636b796368657373120569726f68611a20f697ead9c41f007af0dac2c24ee9881f7d51ea3761e499428fa5523676a15b1c12026d65180320ced99ed10512640a20132f93747b3ad9dd91252b582aa22c5c1510de1d51778e0c9600e420810320161240680647e1baec712491fe9ec9da1e14a32d946cc99ca64d346a069718be67b47dbc99375ffea67a7ac7f28ba36de2afa579decf9c8b42da51c399b84976957c02 ]
Blob: [0a410a331a310a0d66796f646f726b656b2d73616e12205688d35cff09f06f87fae2d7517c8f07c8bba25da667652950a1aa0544c990c812026d65180420cfd99ed10512640a20132f93747b3ad9dd91252b582aa22c5c1510de1d51778e0c9600e4208103201612403278ae4c0dbf58cfe91430ca93262e3d956917b241ec57f99d9750089a5e5f6fbe51a494ca40459faa2aa95ee1f32b4503e564a6e420590840eaa0835c5a2b01 ]
Blob: [0a410a333a310a0d66796f646f726b656b2d73616e12205688d35cff09f06f87fae2d7517c8f07c8bba25da667652950a1aa0544c990c812026d65180520d0d99ed10512640a20132f93747b3ad9dd91252b582aa22c5c1510de1d51778e0c9600e420810320161240365784f245d4f52e98fb0b16627809204f5accda7e4e2fb1f7433e6a009db9b9aca70bf96005933662df44c8fa9c7e441977dee8ccaee4c8345dc6eb38011804 ]
Blob: [0a210a134a110a0d66796f646f726b656b2d73616e100212026d65180620d1d99ed10512640a20132f93747b3ad9dd91252b582aa22c5c1510de1d51778e0c9600e420810320161240fb3c1019dc9ba07a60017694440c19aeabeb304dcce47c58e8b5f7ceeb42079360a32bc0f6829a814cd0d513de2e21491b80779991d3b77198de6316d8b23e0e ]
Blob: [0a3c0a2e0a2c0a106c75636b7963686573734069726f6861120f69726f6861636f696e2369726f68611a070a0320b960100212026d65180720d2d99ed10512640a20132f93747b3ad9dd91252b582aa22c5c1510de1d51778e0c9600e4208103201612400db7e08303f76fcc9ca05f81cfe31f4b82d52373eb5e16a658c07bd0d7ce586ae1354856cd117068d0493eb540e12bc840d207a1cb262d148f6be0d9ee34950e ]
Blob: [0a560a4852460a106c75636b7963686573734069726f6861120b6b616d696c4069726f68611a0f69726f6861636f696e2369726f6861220b68656c6c6f20776f726c642a070a0320dd09100212026d65180820d3d99ed10512640a20132f93747b3ad9dd91252b582aa22c5c1510de1d51778e0c9600e420810320161240335033f765530c1ceef85c4eea5b4ca8eadc11e2848609fb94bbaffe392750cedb9b53f3106bf945d2f20659eb519aad7aa947d49d58fa2f6a07b6d61d85960b ]
```
Also please note that shared object library for Python (_iroha.so) is different from Java library (libiroha.so) even though they provide exactly the same functionality.

**Note about Java:** if you run your Java code you've just built you will probably give the ```java.lang.UnsatisfiedLinkError``` exception. The reason is because Java runtime can't find native .so library (libiroha.so in our case). Just use the next command:
```
LD_LIBRARY_PATH=. java my
```
where **my** is a name of your manually created file.

## P.S. (optional)
This pull request is a kind of a very unfriendly compromise between nice modern C++ code and ability of Swig to interpret it. In some places I was forced to "downgrade" C++ code to make possible to use it. Any ideas about improving it are welcome.

### CODE REVIEW

**Comment by l4l** [comment]
Works fine for python2. But there's a errors with 3rd version :(

File: shared_model/bindings/CMakeLists.txt [hunk 1]
```diff
@@ -18,16 +18,16 @@ if (SWIG_PYTHON OR SWIG_JAVA)
...
 
 if (SWIG_PYTHON)
     find_package(PythonLibs 2.7 REQUIRED)
     # path to where Python.h is found
     include_directories(${PYTHON_INCLUDE_DIRS})
 
-    swig_add_library(irohacrypto LANGUAGE python SOURCES crypto.i)
-    swig_link_libraries(irohacrypto ${Python_LIBRARIES} shared_ed25519_sha3)
+    swig_add_library(irohalib LANGUAGE python SOURCES bindings.i)
+    swig_link_libraries(irohalib ${Python_LIBRARIES} shared_model_proto_backend shared_model_ed25519_sha3 "-flat_namespace -undefined suppress")
```
Comment (l4l) [inline shared_model/bindings/CMakeLists.txt]: This flags cause:
```
clang-5.0: warning: argument unused during compilation: '-flat_namespace' [-Wunused-command-line-argument]
clang-5.0: warning: argument unused during compilation: '-undefined suppress' [-Wunused-command-line-argument]
```
And with -Werror will result error afterwards

File: shared_model/backend/protobuf/transaction.hpp [hunk 2]
```diff
@@ -61,26 +61,39 @@ namespace shared_model {
...
 
-      const CommandsType &commands() const override { return *commands_; }
+      const shared_model::interface::Transaction::CommandsType &commands()
+          const override {
+        return *commands_;
+      }
 
-      const BlobType &blob() const override { return *blob_; }
+      const crypto::Blob &
+      blob() const override {
```
Comment (l4l) [inline shared_model/backend/protobuf/transaction.hpp]: Checked it, wrong formatting here

File: shared_model/builders/protobuf/proto_transaction_builder.hpp [hunk 3]
```diff
@@ -25,38 +25,25 @@
...
-      template <int Sp>
-      TemplateTransactionBuilder(const TemplateTransactionBuilder<Sp> &o)
-          : transaction_(o.transaction_) {}
-
      public:
-      TemplateTransactionBuilder() = default;
+      TransactionBuilder() = default;
 
-      NextBuilder<CreatorAccountId> creatorAccountId(
+      TransactionBuilder creatorAccountId(
```
Comment (l4l) [inline shared_model/builders/protobuf/proto_transaction_builder.hpp]: Something should be done with the swig so it's able to handle proper builder. I don't really like to merge this fixes

File: shared_model/bindings/CMakeLists.txt (hunk 1 above)
Comment (Warchant) [reply shared_model/bindings/CMakeLists.txt]: rename it to `iroha` or `libiroha`

File: shared_model/bindings/simple_builder.cpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "bindings/simple_builder.hpp"
+
+namespace shared_model {
+  namespace proto {
+    SimpleBuilder::SimpleBuilder(const TransactionBuilder &o) : builder_(o) {}
```
Comment (muratovv) [inline shared_model/bindings/simple_builder.cpp]: Maybe move interface for bindings in own separated namespace?

File: shared_model/bindings/simple_builder.hpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+#ifndef IROHA_SIMPLE_BUILDER_HPP
+#define IROHA_SIMPLE_BUILDER_HPP
+
+#include "builders/protobuf/proto_transaction_builder.hpp"
+
+namespace shared_model {
+  namespace proto {
+    class SimpleBuilder {
```
Comment (muratovv) [inline shared_model/bindings/simple_builder.hpp]: I request to add documentation and write a note about this code just for bindings.

File: shared_model/bindings/simple_signer.hpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+#ifndef IROHA_SIMPLE_SIGNER_HPP
+#define IROHA_SIMPLE_SIGNER_HPP
+
+#include "cryptography/blob.hpp"
+#include "cryptography/keypair.hpp"
+#include "cryptography/signed.hpp"
+
+namespace shared_model {
+  namespace proto {
+    class SimpleSigner {
```
Comment (muratovv) [inline shared_model/bindings/simple_signer.hpp]: Same ideas about documentation here.

File: shared_model/bindings/simple_transaction_proto.cpp [hunk 7]
```diff
@@ -21,17 +21,12 @@
...
   namespace proto {
-    bool SimpleTransactionProto::addSignature(Transaction &tx,
-                                              crypto::Signed sign,
-                                              crypto::PublicKey publicKey) {
-      iroha::protocol::Signature protosig;
-      protosig.set_pubkey(publicKey.blob());
-      protosig.set_signature(sign.blob());
-      Signature *s1 = new Signature(protosig);
-      return tx.addSignature(detail::PolymorphicWrapper<Signature>(s1));
+    Transaction SimpleTransactionProto::signAndAddSignature(
```
Comment (l4l) [inline shared_model/bindings/simple_transaction_proto.cpp]: This name is awful

File: shared_model/builders/protobuf/proto_transaction_builder.hpp [hunk 8]
```diff
@@ -65,8 +66,9 @@ namespace shared_model {
         return *this;
       }
 
-      Transaction build() {
-        return Transaction(std::move(transaction_));
+      shared_model::proto::UnsignedWrapper<Transaction> build() {
```
Comment (l4l) [inline shared_model/builders/protobuf/proto_transaction_builder.hpp]: redundant namespace prefix

File: shared_model/builders/protobuf/unsigned_proto.hpp [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+#include "cryptography/keypair.hpp"
+
+namespace shared_model {
+  namespace proto {
+    /**
+     * Class for holding built but still unsigned objects
+     * @tparam T - type of object received from builder
+     */
+    template <typename T>
+    class UnsignedWrapper {
```
Comment (l4l) [inline shared_model/builders/protobuf/unsigned_proto.hpp]: Awful name[2]
Maybe change it to `Incomplete`?

File: test/module/shared_model/backend_proto/shared_proto_transaction_test.cpp [hunk 10]
```diff
@@ -52,12 +55,22 @@ TEST(ProtoTransaction, Builder) {
...
 
-  auto tx = shared_model::proto::TransactionBuilder()
-                .txCounter(tx_counter)
-                .creatorAccountId(account_id)
-                .addAssetQuantity(account_id, asset_id, amount)
-                .build();
-  auto &proto = tx.getTransport();
+  auto keypair =
+      shared_model::crypto::CryptoProviderEd25519Sha3::generateKeypair();
+  auto signedProto = shared_model::crypto::CryptoSigner<>::sign(
```
Comment (l4l) [inline test/module/shared_model/backend_proto/shared_proto_transaction_test.cpp]: Split this test to signed & unsigned protos

Comment (luckychess) [reply test/module/shared_model/backend_proto/shared_proto_transaction_test.cpp]: Unsigned protos are "illegal" and should not be available via API so test for unsigned will be pretty useless.

File: shared_model/bindings/bindings.i [hunk 11]
```diff
@@ -15,47 +15,55 @@
...
+%include "bindings/simple_transaction_proto.hpp"
+
+%template (UnsignedTx) shared_model::proto::UnsignedWrapper<shared_model::proto::Transaction>;
 
 namespace shared_model {
-  namespace interface {
-    %template(imodelprim) shared_model::interface::ModelPrimitive<shared_model::crypto::Blob>;
-    %template(ikeypair) ModelPrimitive<crypto::Keypair>;
-    %template(iprim) shared_model::interface::Primitive<crypto::Keypair, iroha::keypair_t>;
+  namespace proto {
```
Comment (muratovv) [inline shared_model/bindings/bindings.i]: Should it be renamed with bindings?

File: shared_model/bindings/simple_transaction_proto.hpp [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    class SimpleTransactionProto {
+     public:
+      /**
+       * Signs unsigned transaction and adds signature to its internal proto
+       * object
+       * @param tx - unsigned transaction
+       * @param keypair - keypair to sign
+       * @return signed transaction
+       */
+      proto::Transaction signAndAddSignature(
```
Comment (muratovv) [inline shared_model/bindings/simple_transaction_proto.hpp]: Better to change it with `Blob`.

File: shared_model/bindings/model_crypto.cpp [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+#include "bindings/model_crypto.hpp"
+#include "cryptography/ed25519_sha3_impl/crypto_provider.hpp"
+#include "cryptography/seed.hpp"
+
+namespace shared_model {
+  namespace bindings {
+    crypto::Keypair ModelCrypto::generateKeypair() {
+      return crypto::CryptoProviderEd25519Sha3::generateKeypair();
```
Comment (muratovv) [inline shared_model/bindings/model_crypto.cpp]: Add TODO for fixing it when we have the default interface for key pair generation.

File: shared_model/builders/protobuf/transaction.hpp [hunk 14]
```diff
@@ -163,14 +164,17 @@ namespace shared_model {
         return *this;
       }
 
-      Transaction build() {
+      UnsignedWrapper<Transaction> build() {
         static_assert(S == (1 << TOTAL) - 1, "Required fields are not set");
 
-        return Transaction(iroha::protocol::Transaction(transaction_));
+        return UnsignedWrapper<Transaction>(
+            Transaction(std::move(transaction_)));
```
Comment (muratovv) [inline shared_model/builders/protobuf/transaction.hpp]: Rework with copy

**Pull request merged**